src/main/java/htsjdk/samtools/cram/encoding/reader/CramRecordReader.java

```diff
@@ -20,33 +20,164 @@
+        hardClipCodec =                 createDataReader(DataSeries.HC_HardClip);
+        paddingCodec =                  createDataReader(DataSeries.PD_padding);
+        deletionLengthCodec =           createDataReader(DataSeries.DL_DeletionLength);
+        mappingScoreCodec =             createDataReader(DataSeries.MQ_MappingQualityScore);
+        mateBitFlagCodec =              createDataReader(DataSeries.MF_MateBitFlags);
+        mateReferenceIdCodec =          createDataReader(DataSeries.NS_NextFragmentReferenceSequenceID);
+        mateAlignmentStartCodec =       createDataReader(DataSeries.NP_NextFragmentAlignmentStart);
+        insertSizeCodec =               createDataReader(DataSeries.TS_InsetSize);
+        tagIdListCodec =                createDataReader(DataSeries.TL_TagIdList);
+        refIdCodec =                    createDataReader(DataSeries.RI_RefId);
+        refSkipCodec =                  createDataReader(DataSeries.RS_RefSkip);
+        basesCodec =                    createDataReader(DataSeries.BB_bases);
+        scoresCodec =                   createDataReader(DataSeries.QQ_scores);
+
+        // special case: re-encodes QS as a byte array
+        qualityScoreArrayCodec = new DataSeriesReader<>(DataSeriesType.BYTE_ARRAY, header.encodingMap.get(DataSeries.QS_QualityScore), coreInputStream, externalInputMap);
+
+        tagValueCodecs = header.tMap.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey,
+                        mapEntry -> new DataSeriesReader<>(DataSeriesType.BYTE_ARRAY, mapEntry.getValue(), coreInputStream, externalInputMap)));
+    }
+
+    /**
+     * Initialize a Cram Record Reader from a Slice
+     *
+     * @param slice the associated Slice
+     * @param header the associated Cram Compression Header
+     * @param validationStringency how strict to be when reading this CRAM record
+     */
+    public CramRecordReader(final Slice slice,
+                            final CompressionHeader header,
+                            final ValidationStringency validationStringency) {
+        this(slice.getCoreInputStream(),
+                slice.getExternalInputMap(),
+                header,
+                slice.sequenceId,
+                validationStringency);
+    }
+
+    /**
+     * Look up a Data Series in the Cram Compression Header's Encoding Map.  If found, create a Data Reader
+     *
+     * @param dataSeries Which Data Series to write
+     * @param <T> The Java data type associated with the Data Series
+     * @return a Data Reader for the given Data Series, or null if it's not in the encoding map
+     */
+    private <T> DataSeriesReader<T> createDataReader(DataSeries dataSeries) {
+        if (encodingMap.containsKey(dataSeries)) {
+            return new DataSeriesReader<>(dataSeries.getType(), encodingMap.get(dataSeries), coreBlockInputStream, externalBlockInputMap);
+        } else {
+            return null;
```

Comment:
TODO!  Investigate why the encodingMap might not have a given dataSeries.  We should be able to guarantee this.